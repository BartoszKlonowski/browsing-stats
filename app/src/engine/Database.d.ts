export default class Database {
    writeTimeSpent(domain: string, timeSpentInSeconds: number): void;

    readTimeSpent(onResult: (content: number | Map<string, number>) => void, domain?: string): void;

    writePreviousDomain(domain: string): void;

    readPreviousDomain(onResult: (domain: string) => void): void;

    writeLastActive(domain: string, date: Date): void;

    readLastActive(domain: string, onReturn: (date: Date) => void): void;
}
