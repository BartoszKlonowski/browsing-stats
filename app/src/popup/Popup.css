.popup-view {
  border: 5px solid #00e080;
  background-color: #F2F2F2;
  align-content: space-around;
  padding: 2px;
}
.view-change-button {
  display: flex;
  padding: 3px;
  border: 3px solid #FC4D17;
  border-radius: 10px;
  flex-direction: row;
  justify-items: center;
  justify-content: center;
}
.view-change-button-text {
  font-family: Georgia, 'Times New Roman', Times, serif;
  font-size: small;
  padding: 1px;
}
.duration-header {
  display: flex;
  flex-direction: row;
  align-content: center;
  justify-content: space-between;
  padding: 5px;
  height: 40px;
}
.duration-header-text {
  font-family: Georgia, 'Times New Roman', Times, serif;
  font-size: medium;
  padding: 5px;
}
.sorting-container {
  padding: 5px;
  display: flex;
  flex-direction: row;
  width: 240px;
  cursor: default;
}
.sorting-option-tile-container {
  margin: 2px;
  cursor: pointer;
}
.sorting-option-tile-label {
  font-family: Georgia, 'Times New Roman', Times, serif;
  font-size: small;
}
.sorting-options-label {
  font-family: Georgia, 'Times New Roman', Times, serif;
  padding: 2px;
  font-size: small;
}
.sorting-options-list-container {
  height: fit-content;
  position: relative;
  background-color: #F2F2F2;
  box-shadow: 0px 3px 3px #FC4D17;
  border-left: 2px solid #FC4D17;
  border-radius: 5px;
}
.shrinked-view-container {
  padding: 5px;
  height: 40px;
  width: 120px;
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  align-items: center;
}
.shrinked-view-icon {
  position: relative;
}
.shrinked-view-time-spent-text {
  font-size: medium;
  text-align: center;
  position: relative;
  align-content: center;
  align-self: center;
  align-items: center;
}
.shrinked-view-time-spent-text-container {
  height: 40px;
  width: 70px;
  display: flex;
  text-align: center;
  justify-content: center;
  align-items: center;
}
.shrinked-view-icon-container {
  height: 40px;
  width: 40px;
  display: flex;
  justify-content: center;
  align-items: center;
}
.expanded-view-list {
  max-height: 300px;
  overflow-y: auto;
  overflow-x: hidden;
  list-style: none;
  scrollbar-width: thin;
  scrollbar-color: #FC4D17 transparent;
}
.expanded-view-list-container {
  max-width: 500px;
  padding: 5px;
  overflow: hidden;
  scrollbar-color: #FC4D17;
}
.expanded-view-tile-container {
  display: flex;
  flex-direction: row;
  justify-content: left;
  padding: 5px;
  height: 45px;
}
.expanded-view-icon-container {
  padding-left: 5px;
  padding-right: 5px;
}
.expanded-view-icon {
  width: 24px;
  height: 24px;
}
.expanded-view-domain-container {
  width: 300px;
  padding-left: 5px;
  padding-right: 5px;
}
.expanded-view-domain-text {
  font-family: Georgia, 'Times New Roman', Times, serif;
  font-size: medium;
  cursor: pointer;
  text-decoration: none;
  background-image: linear-gradient(to right, #FC4D17, #FC4D17);
  background-position: bottom left;
  background-repeat: no-repeat;
  background-size: 10% 1px;
  transition: background-size .5s ease;
}
.expanded-view-domain-text:hover {
  background-size: 50% 1px;
}
.expanded-view-time-spent-text-container {
  width: 100px;
  padding-left: 5px;
  padding-right: 5px;
}
.expanded-view-time-spent-text {
  font-size: medium;
  font-weight: bold;
}
.expanded-view-tile-details-navigation-button {
  font-family: Georgia, 'Times New Roman', Times, serif;
  font-size: small;
  max-width: max-content;
  padding: 3px;
  cursor: pointer;
  box-shadow: 0px 3px 3px #212121;
}
.expanded-view-tile-details-navigation-button:hover {
  border-radius: 5px;
  color: #FC4D17;
  transition: .2s ease;
}
.details-view-container {
  height: 360px;
  width: 500px;
  padding: 5px;
}
.details-view-website-title {
  padding: 5px;
  font-family: Georgia, 'Times New Roman', Times, serif;
  font-size: medium;
  font-weight: bold;
  color: #FC4D17;
}
.details-view-details-table {
  margin-top: 10px;
  margin-bottom: 50px;
  display: flex;
  flex-direction: column;
}
.details-view-table-row {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 100%;
  border-bottom-style: ridge;
  border-bottom-width: 1px;
  border-bottom-color: #FC4D17;
}
.details-view-table-label {
  padding: 5px;
  justify-items: flex-end;
  justify-self: flex-end;
  font-family: Georgia, 'Times New Roman', Times, serif;
  font-size: medium;
}
.details-view-table-value {
  padding: 5px;
  font-weight: bold;
  font-size: medium;
}
.details-view-back-button {
  font-family: Georgia, 'Times New Roman', Times, serif;
  font-size: medium;
  max-width: max-content;
  margin: 5px;
  padding: 3px;
  cursor: pointer;
  box-shadow: 0px 3px 3px #212121;
}
.details-view-back-button:hover {
  border-radius: 5px;
  color: #FC4D17;
  transition: .2s ease;
}
