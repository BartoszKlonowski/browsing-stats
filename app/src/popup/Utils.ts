import Database from "../engine/Database";

export interface Icon {
    size: number;
    src: string;
}

export enum Sort {
    NameAscending = "option-nameascending",
    NameDescending = "option-namedescending",
    TimeAscending = "option-timeascending",
    TimeDescending = "option-timedescending",
    None = "option-none",
}

export const optionEventValueToSortEnum = (optionEvent?: string): Sort => {
    switch (optionEvent) {
        case Sort.NameAscending:
        case Sort.NameDescending:
        case Sort.TimeAscending:
        case Sort.TimeDescending:
            return optionEvent as Sort;
        default:
            return Sort.None;
    }
};

export const getActiveTabDomainFromURL = (URL: string): string | null => {
    let result = URL.replace("https://", "");
    result = result.replace("http://", "");
    result = result.replace("www.", "");
    const results = result.match(/[^ /]+/g);
    return results && !results[0].includes(" ") && results[0].includes(".") ? results[0] : null;
};

export const getWebsiteIconObject = (websiteURL: string | undefined): Icon => {
    const iconDesiredSize = 20;
    const iconSource: string =
        websiteURL && websiteURL.length
            ? `https://icons.duckduckgo.com/ip3/${websiteURL}.ico`
            : "../resources/missing-website-favicon.png";

    const icon: Icon = {
        size: iconDesiredSize,
        src: iconSource,
    };
    return icon;
};

export function storeTimeSpentSummary(currentDomain: string) {
    const db = new Database();
    db.readPreviousDomain((domain: string) => {
        if (!!currentDomain.length && domain !== currentDomain) {
            db.writePreviousDomain(currentDomain);
            db.writeLastActive(currentDomain, new Date());
            calculateTimeSpentForDomain(domain);
        }
    });
}

export const sortDataEntries = (data: Map<string, number>, sortMethod: Sort) => {
    return new Map(
        [...data].sort((entry1: [string, number], entry2: [string, number]) => {
            switch (sortMethod) {
                case Sort.NameAscending:
                    return entry1[0] < entry2[0] ? -1 : 1;
                case Sort.NameDescending:
                    return entry1[0] > entry2[0] ? -1 : 1;
                case Sort.TimeAscending:
                    return entry1[1] - entry2[1];
                case Sort.TimeDescending:
                    return entry2[1] - entry1[1];
                default:
                    return 0;
            }
        })
    );
};

export function calculateTimeSpentForDomain(domain: string) {
    const db = new Database();
    db.readLastActive(domain, (date: Date) => {
        db.readTimeSpent((content: number | Map<string, number>) => {
            db.writeTimeSpent(domain, Math.trunc(Math.abs(Date.now() - date.getTime()) / 1000) + (content as number));
        }, domain);
    });
}

export const howManyHoursInSeconds = (timeInSeconds: number): number => {
    const secondsInOneHour = 60 * 60;
    return (timeInSeconds - (timeInSeconds % secondsInOneHour)) / secondsInOneHour;
};

export const howManyMinutesInSeconds = (timeInSeconds: number): number => {
    const secondsInOneMinute = 60;
    if (timeInSeconds > secondsInOneMinute) {
        return Math.trunc((timeInSeconds / secondsInOneMinute) % secondsInOneMinute);
    } else {
        return 0;
    }
};

export const howManySecondsInSeconds = (timeInSeconds: number): number => {
    const secondsInOneMinute = 60;
    return timeInSeconds % secondsInOneMinute;
};

export const getMinutes = (timeInSeconds: number) => {
    const minutes = howManyMinutesInSeconds(timeInSeconds);
    return minutes < 10 ? `0${minutes}` : minutes.toString();
};

export const getSeconds = (timeInSeconds: number) => {
    const seconds = howManySecondsInSeconds(timeInSeconds);
    return seconds < 10 ? `0${seconds}` : seconds.toString();
};

export const getHours = (timeInSeconds: number) => howManyHoursInSeconds(timeInSeconds);
