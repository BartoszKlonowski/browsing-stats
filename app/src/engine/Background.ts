import browser from "webextension-polyfill";
import {
    calculateTimeSpentForDomain,
    getActiveTabDomainFromURL,
    storeFirstVisitDateForDomain,
    storeVisitsNumber,
} from "../popup/Utils";
import {storeTimeSpentSummary} from "../popup/Utils";
import Database from "./Database";

browser.tabs.onActivated.addListener((activeInfo) => {
    browser.tabs.get(activeInfo.tabId).then((tab) => {
        storeTimeSpentSummary(getActiveTabDomainFromURL(tab?.url || "") || "");
    });
});

browser.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
    if (changeInfo.status == "complete") {
        const activeDomain = getActiveTabDomainFromURL(tab.url || "");
        storeFirstVisitDateForDomain(activeDomain || "");
        storeVisitsNumber(activeDomain);
        storeTimeSpentSummary(activeDomain || "");
    }
});

browser.windows.onFocusChanged.addListener((id) => {
    const windowInactiveID = -1;
    const db = new Database();
    if (id === windowInactiveID) {
        db.readPreviousDomain((domain: string) => {
            calculateTimeSpentForDomain(domain);
            db.writePreviousDomain("");
        });
    } else {
        browser.tabs.query({active: true}).then((tab) => {
            const newFocusedDomain = getActiveTabDomainFromURL(tab[0].url || "");
            db.writeLastActive(newFocusedDomain || "", new Date());
            db.writePreviousDomain(newFocusedDomain || "");
        });
    }
});
