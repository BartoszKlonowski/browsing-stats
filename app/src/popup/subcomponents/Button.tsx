import React from "react";
import {translate} from "../../engine/i18n";

interface Props {
    label: string;
    onClick: () => void;
}

export const Button = ({label, onClick}: Props): JSX.Element => {
    return (
        <button
            className="button"
            itemType="button"
            onClick={onClick}>
            <div className="button-text">
                {label}
            </div>
        </button>
    );
};

export default Button;
