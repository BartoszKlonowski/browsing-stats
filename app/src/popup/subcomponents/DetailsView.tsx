import React, {useMemo, useState} from "react";
import Database from "../../engine/Database";
import {translate} from "../../engine/i18n";

interface Props {
    website: string;
    onBackButtonClick: () => void;
}

const DetailsView = ({website, onBackButtonClick}: Props): React.JSX.Element => {
    const [lastVisited, setLastVisited] = useState<string>("...");
    const db = new Database();

    useMemo(() => {
        db.readLastVisited(website, (date) => {
            setLastVisited(date.toLocaleString());
        });
    }, []);

    const details = [{detail: translate("details-view-last-visited-label"), value: lastVisited}];

    return (
        <div className="details-view-container">
            <div className="details-view-website-title">{website}</div>
            <div className="details-view-details-table">
                {details.map((row) => (
                    <div className="details-view-table-row" key={row.detail}>
                        <div className="details-view-table-label">{`${row.detail}:`}</div>
                        <div className="details-view-table-value">{row.value}</div>
                    </div>
                ))}
            </div>
            <div className="details-view-back-button" onClick={onBackButtonClick}>
                {translate("details-view-back-button-label")}
            </div>
        </div>
    );
};

export default DetailsView;
