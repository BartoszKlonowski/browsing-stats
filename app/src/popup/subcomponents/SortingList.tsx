import React, {useState} from "react";
import {translate} from "../../engine/i18n";
import {Sort} from "../Utils";

interface Props {
    onSelect: (option: Sort) => void;
}

interface OptionProps {
    label: string;
    selected: () => void;
}

interface OptionsListProps {
    options: {value: Sort; label: string}[];
    onSelected: (value: Sort) => void;
}

const Option = ({label, selected}: OptionProps) => {
    return (
        <div
            className="sorting-option-tile-container"
            onClick={() => {
                selected();
            }}>
            <div className="sorting-option-tile-label">{label}</div>
        </div>
    );
};

const OptionsList = ({options, onSelected}: OptionsListProps) => {
    return (
        <div className="sorting-options-list-container">
            {options.map((option) => {
                return (
                    <Option
                        key={`option-${option.value}`}
                        label={option.label}
                        selected={() => onSelected(option.value)}
                    />
                );
            })}
        </div>
    );
};

const SortingList = ({onSelect}: Props) => {
    const [currentOption, setCurrentOption] = useState(Sort.None);
    const [isExpanded, setExpand] = useState(false);
    const options = [
        {value: Sort.None, label: translate(Sort.None)},
        {value: Sort.NameAscending, label: translate(Sort.NameAscending)},
        {value: Sort.NameDescending, label: translate(Sort.NameDescending)},
        {value: Sort.TimeAscending, label: translate(Sort.TimeAscending)},
        {value: Sort.TimeDescending, label: translate(Sort.TimeDescending)},
    ];
    return (
        <div className="sorting-container">
            <div className="sorting-options-label">{`${translate("sort-label")}: `}</div>
            <div className="sorting-header-button" onClick={() => setExpand(!isExpanded)}>
                {`${options.find((option) => option.value === currentOption)?.label}`}
            </div>
            {isExpanded ? (
                <OptionsList
                    options={options}
                    onSelected={(value) => {
                        setCurrentOption(value);
                        onSelect(value);
                        setExpand(!isExpanded);
                    }}
                />
            ) : null}
        </div>
    );
};

export default SortingList;
