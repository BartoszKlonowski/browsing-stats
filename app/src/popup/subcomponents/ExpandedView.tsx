import React, {useState} from "react";
import {Sort} from "../Utils";
import DetailsView from "./DetailsView";
import DurationHeader from "./DurationHeader";
import TimeSpentList from "./TimeSpentList";

interface Props {
    sorted: Sort;
    setSortingOrder: (order: Sort) => void;
}

export const ExpandedView = ({sorted, setSortingOrder}: Props) => {
    const [detailsScreenWebsite, setDetailsScreenWebsite] = useState<string | null>();

    return (
        <>
            {!detailsScreenWebsite ? <DurationHeader onSortSelected={setSortingOrder} /> : null}
            {detailsScreenWebsite ? (
                <DetailsView website={detailsScreenWebsite} />
            ) : (
                <div className="expanded-view-list-container">
                    <TimeSpentList sorted={sorted} onEnterClick={(domain) => setDetailsScreenWebsite(domain)} />
                </div>
            )}
        </>
    );
};

export default ExpandedView;
