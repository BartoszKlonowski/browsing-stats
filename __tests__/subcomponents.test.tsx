import React from "react";
import TestRenderer, {act, ReactTestInstance, ReactTestRenderer} from "react-test-renderer";
import {fireEvent, render, screen} from "@testing-library/react";
import ShrinkedView from "../app/src/popup/subcomponents/ShrinkedView";
import ExpandedView from "../app/src/popup/subcomponents/ExpandedView";
import Button from "../app/src/popup/subcomponents/Button";
import SortingList from "../app/src/popup/subcomponents/SortingList";
import DetailsView from "../app/src/popup/subcomponents/DetailsView";
import {Sort} from "../app/src/popup/Utils";
import {translate} from "../app/src/engine/i18n";

function renderElement(element: JSX.Element): ReactTestRenderer {
    const component = TestRenderer.create(element);
    return component;
}

async function renderElementAsObject(element: JSX.Element): Promise<ReactTestInstance> {
    let component;
    await act(async () => {
        component = TestRenderer.create(element);
    });
    return JSON.parse(JSON.stringify(component.toJSON()));
}

function getChild(renderedObject: ReactTestInstance, childIndex: number): ReactTestInstance {
    const child = renderedObject.children[childIndex] as ReactTestInstance;
    return child;
}

beforeEach(() => {
    global.browser.i18n.getUILanguage = () => "EN";
});

describe("ShrinkedView", () => {
    it("renders correctly according to snapshot", () => {
        const shrinkedView = renderElement(<ShrinkedView />);
        expect(shrinkedView.toJSON()).toMatchSnapshot();
    });

    it("contains the duration header at the beginning of DOM", async () => {
        global.browser.tabs.query = () => {
            return new Promise((resolve) => {
                resolve([{url: ""}]);
            });
        };
        const shrinkedView = await renderElementAsObject(<ShrinkedView />);
        expect(shrinkedView[0].props.className).toBe("duration-header");
    });

    it("renders icon as image with default src when missing icon on website", async () => {
        global.browser.tabs.query = () => {
            return new Promise((resolve) => {
                resolve([{url: ""}]);
            });
        };
        const shrinkedView = await renderElementAsObject(<ShrinkedView />);
        const img = getChild(getChild(shrinkedView[1], 0), 0);
        expect(img).toBeDefined();
        expect(img.type).toBe("img");
        expect(img.props.src).toContain("../resources/missing-website-favicon.png");
    });

    it("renders icon as image with proper src when icon is available on website", async () => {
        global.browser.tabs.query = () => {
            return new Promise((resolve) => {
                resolve([{url: "proper.existing.icon.url"}]);
            });
        };
        const shrinkedView = await renderElementAsObject(<ShrinkedView />);
        const img = getChild(getChild(shrinkedView[1], 0), 0);
        expect(img).toBeDefined();
        expect(img.type).toBe("img");
        expect(img.props.src).toContain("https://icons.duckduckgo.com/ip3/proper.existing.icon.url.ico");
    });

    it("displays 0 time spent for incorrect domain", async () => {
        global.browser.tabs.query = () => {
            return new Promise((resolve) => {
                resolve([{url: ""}]);
            });
        };
        const shrinkedView = await renderElementAsObject(<ShrinkedView />);
        const timeSpentText = getChild(getChild(shrinkedView[1], 1), 0);
        expect(timeSpentText).toBeDefined();
        expect(`${getChild(timeSpentText, 0)}:${getChild(timeSpentText, 2)}:${getChild(timeSpentText, 4)}`).toBe(
            "0:00:00"
        );
    });
});

describe("ExpandedView", () => {
    it("renders correctly according to snapshot", () => {
        const expandedView = renderElement(
            <ExpandedView
                setSortingOrder={(order) => {
                    order;
                }}
                sorted={Sort.None}
            />
        );
        expect(expandedView.toJSON()).toMatchSnapshot();
    });

    it("contains the duration header at the beginning of DOM", async () => {
        global._localStorage.getItem = (key: string) => {
            return key ? `{}` : null;
        };
        const expandedView = await renderElementAsObject(
            <ExpandedView
                setSortingOrder={(order) => {
                    order;
                }}
                sorted={Sort.None}
            />
        );
        expect(expandedView[0].props.className).toBe("duration-header");
    });

    it("contains an empty list when no data in timeSpent storage", async () => {
        global._localStorage.getItem = (key: string) => {
            return key ? `{}` : null;
        };
        const expandedView = await renderElementAsObject(
            <ExpandedView
                setSortingOrder={(order) => {
                    order;
                }}
                sorted={Sort.None}
            />
        );
        expect(expandedView[1]).toBeDefined();
        expect(getChild(expandedView[1], 0).children).toBeNull();
    });

    it("contains a list items of proper type and layout", async () => {
        global._localStorage.getItem = (key: string) => {
            return `[["${key}", 20], ["fake", 20]]`;
        };
        const expandedView = await renderElementAsObject(
            <ExpandedView
                setSortingOrder={(order) => {
                    order;
                }}
                sorted={Sort.None}
            />
        );
        const mainList = getChild(expandedView[1], 0);
        expect(mainList.type).toBe("ul");
        expect(getChild(mainList, 0).type).toBe("li");
        expect(getChild(mainList, 1).type).toBe("li");
    });

    it("contains a list with proper number of items", async () => {
        global._localStorage.getItem = (key: string) => {
            return `[["first item", 10],["second item", 20],["${key}", 30]]`;
        };
        const expandedView = await renderElementAsObject(
            <ExpandedView
                setSortingOrder={(order) => {
                    order;
                }}
                sorted={Sort.None}
            />
        );
        expect(getChild(expandedView[1], 0).children.length).toBe(3);
    });
});

describe("Button", () => {
    it("renders correctly regarding the snapshot", () => {
        const button = renderElement(
            <Button
                label={"test"}
                onClick={() => {}}
            />
        );
        expect(button).toMatchSnapshot();
    });

    it("is created by a button type", async () => {
        const button = await renderElementAsObject(
            <Button
                label={"test"}
                onClick={() => {}}
            />
        );
        expect(button.type).toBe("button");
    });
});

describe("Sorting", () => {
    it("renders correctly according to snapshot", () => {
        const sortList = renderElement(
            <SortingList
                onSelect={(_: Sort): void => {
                    _;
                }}
            />
        );
        expect(sortList).toMatchSnapshot();
    });

    it("initially renders with the option 'None'", async () => {
        const sortList = await renderElementAsObject(
            <SortingList
                onSelect={(_: Sort): void => {
                    _;
                }}
            />
        );
        expect(getChild(getChild(getChild(getChild(sortList, 1), 0), 0), 0)).toBe("None");
    });

    it("does not render the list when not clicked", async () => {
        const sortList = await renderElementAsObject(
            <SortingList
                onSelect={(_: Sort): void => {
                    _;
                }}
            />
        );
        expect(getChild(sortList, 2)).toBeUndefined();
    });

    it("renders the complete list when clicked", async () => {
        const sortList = await renderElementAsObject(
            <SortingList
                onSelect={(_: Sort): void => {
                    _;
                }}
            />
        );
        expect(getChild(sortList, 1)).toBeDefined();
    });
});

describe("DetailsView", () => {
    it("renders correctly according to snapshot", () => {
        global._localStorage.getItem = (key: string) => {
            return `[["test", "invalid-mocked-date"],["${key}", "${new Date(0)}"]]`;
        };
        const detailsView = renderElement(<DetailsView website={"test"} onBackButtonClick={() => undefined} />);
        expect(detailsView).toMatchSnapshot();
    });

    it("contains the back button for back navigation", async () => {
        const detailsView = await renderElementAsObject(
            <DetailsView website="test" onBackButtonClick={() => undefined} />
        );
        expect(getChild(detailsView, 2).props.className).toBe("button");
    });

    it("calls the onBackButton handler when back button is clicked", async () => {
        const mockedBackButtonHandler = jest.fn();
        render(<DetailsView website="test" onBackButtonClick={mockedBackButtonHandler} />);
        const backButton = screen.getByText("Go back");
        fireEvent.click(backButton);
        expect(mockedBackButtonHandler).toBeCalled();
    });

    it("contains the last visited with proper label and value matching storage state", async () => {
        const currentDate = new Date();
        global._localStorage.setItem = jest.fn();
        global._localStorage.getItem = (key: string) => {
            return `[["${key}", "${currentDate}"],["second.domain", "${currentDate}"]]`;
        };
        render(<DetailsView website="test" onBackButtonClick={() => undefined} />);
        const lastVisitedLabel = await screen.findByText(`${translate("details-view-last-visited-label")}:`);
        const lastVisitedValue = await screen.findByText(currentDate.toLocaleString());
        expect(lastVisitedLabel?.textContent).toBe(`${translate("details-view-last-visited-label")}:`);
        expect(lastVisitedValue?.textContent).toBe(currentDate.toLocaleString());
    });

    it("contains the time spent summary with proper label and value matching storage state", async () => {
        global._localStorage.setItem = jest.fn();
        global._localStorage.getItem = (key: string) => {
            return `[["${key}", "120"],["test", "243"]]`;
        };
        render(<DetailsView website="test" onBackButtonClick={() => undefined} />);
        const timeSpentLabel = await screen.findByText(`${translate("duration-header")}:`);
        const timeSpentValue = await screen.findByText("0:04:03");
        expect(timeSpentLabel?.textContent).toBe(`${translate("duration-header")}:`);
        expect(timeSpentValue?.textContent).toBe("0:04:03");
    });

    it("contains the total number of visits with proper label and value matching storage state", async () => {
        global._localStorage.setItem = jest.fn();
        global._localStorage.getItem = (key: string) => {
            return `[["${key}", "120"],["test", "24"]]`;
        };
        render(<DetailsView website="test" onBackButtonClick={() => undefined} />);
        const totalNumberLabel = await screen.findByText(`${translate("details-view-number-of-visits")}:`);
        const totalNumberValue = await screen.findByText("24");
        expect(totalNumberLabel?.textContent).toBe(`${translate("details-view-number-of-visits")}:`);
        expect(totalNumberValue?.textContent).toBe("24");
    });
});
