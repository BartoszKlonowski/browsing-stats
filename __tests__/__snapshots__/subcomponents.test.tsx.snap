// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsView renders correctly according to snapshot 1`] = `
<div
  className="details-view-container"
>
  <div
    className="details-view-website-title"
  >
    test
  </div>
  <div
    className="details-view-details-table"
  >
    <div
      className="details-view-table-row"
    >
      <div
        className="details-view-table-label"
      >
        Last visit:
      </div>
      <div
        className="details-view-table-value"
      >
        Invalid Date
      </div>
    </div>
  </div>
  <div
    className="details-view-back-button"
    onClick={[Function]}
  >
    Go back
  </div>
</div>
`;

exports[`ExpandedView renders correctly according to snapshot 1`] = `
Array [
  <div
    className="duration-header"
  >
    <div
      className="duration-header-text"
    >
      Spent here
    </div>
    <div
      className="sorting-container"
    >
      <div
        className="sorting-options-label"
      >
        Sort: 
      </div>
      <div
        className="sorting-options-list-container"
      >
        <div
          className="sorting-option-tile-container"
          onClick={[Function]}
        >
          <div
            className="sorting-option-tile-label"
          >
            None
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="expanded-view-list-container"
  >
    <ul
      className="expanded-view-list"
    >
      <li>
        <div
          className="expanded-view-tile-container"
        >
          <div
            className="expanded-view-icon-container"
          >
            <img
              className="expanded-view-icon"
              height={20}
              src="https://icons.duckduckgo.com/ip3/w3schools.com.ico"
              width={20}
            />
          </div>
          <div
            className="expanded-view-domain-container"
          >
            <div
              className="expanded-view-domain-text"
              onClick={[Function]}
            >
              w3schools.com
            </div>
          </div>
          <div
            className="expanded-view-time-spent-text-container"
          >
            <div
              className="expanded-view-time-spent-text"
            >
              0
              :
              01
              :
              13
            </div>
          </div>
          <div
            className="expanded-view-tile-details-navigation-button"
            onClick={[Function]}
          >
            Details
          </div>
        </div>
      </li>
      <li>
        <div
          className="expanded-view-tile-container"
        >
          <div
            className="expanded-view-icon-container"
          >
            <img
              className="expanded-view-icon"
              height={20}
              src="https://icons.duckduckgo.com/ip3/ackoverflow.com.ico"
              width={20}
            />
          </div>
          <div
            className="expanded-view-domain-container"
          >
            <div
              className="expanded-view-domain-text"
              onClick={[Function]}
            >
              ackoverflow.com
            </div>
          </div>
          <div
            className="expanded-view-time-spent-text-container"
          >
            <div
              className="expanded-view-time-spent-text"
            >
              0
              :
              11
              :
              14
            </div>
          </div>
          <div
            className="expanded-view-tile-details-navigation-button"
            onClick={[Function]}
          >
            Details
          </div>
        </div>
      </li>
      <li>
        <div
          className="expanded-view-tile-container"
        >
          <div
            className="expanded-view-icon-container"
          >
            <img
              className="expanded-view-icon"
              height={20}
              src="https://icons.duckduckgo.com/ip3/github.com.ico"
              width={20}
            />
          </div>
          <div
            className="expanded-view-domain-container"
          >
            <div
              className="expanded-view-domain-text"
              onClick={[Function]}
            >
              github.com
            </div>
          </div>
          <div
            className="expanded-view-time-spent-text-container"
          >
            <div
              className="expanded-view-time-spent-text"
            >
              0
              :
              10
              :
              07
            </div>
          </div>
          <div
            className="expanded-view-tile-details-navigation-button"
            onClick={[Function]}
          >
            Details
          </div>
        </div>
      </li>
      <li>
        <div
          className="expanded-view-tile-container"
        >
          <div
            className="expanded-view-icon-container"
          >
            <img
              className="expanded-view-icon"
              height={20}
              src="https://icons.duckduckgo.com/ip3/google.com.ico"
              width={20}
            />
          </div>
          <div
            className="expanded-view-domain-container"
          >
            <div
              className="expanded-view-domain-text"
              onClick={[Function]}
            >
              google.com
            </div>
          </div>
          <div
            className="expanded-view-time-spent-text-container"
          >
            <div
              className="expanded-view-time-spent-text"
            >
              0
              :
              00
              :
              07
            </div>
          </div>
          <div
            className="expanded-view-tile-details-navigation-button"
            onClick={[Function]}
          >
            Details
          </div>
        </div>
      </li>
      <li>
        <div
          className="expanded-view-tile-container"
        >
          <div
            className="expanded-view-icon-container"
          >
            <img
              className="expanded-view-icon"
              height={20}
              src="https://icons.duckduckgo.com/ip3/npmjs.com.ico"
              width={20}
            />
          </div>
          <div
            className="expanded-view-domain-container"
          >
            <div
              className="expanded-view-domain-text"
              onClick={[Function]}
            >
              npmjs.com
            </div>
          </div>
          <div
            className="expanded-view-time-spent-text-container"
          >
            <div
              className="expanded-view-time-spent-text"
            >
              0
              :
              00
              :
              05
            </div>
          </div>
          <div
            className="expanded-view-tile-details-navigation-button"
            onClick={[Function]}
          >
            Details
          </div>
        </div>
      </li>
    </ul>
  </div>,
]
`;

exports[`ShrinkedView renders correctly according to snapshot 1`] = `
Array [
  <div
    className="duration-header"
  >
    <div
      className="duration-header-text"
    >
      Spent here
    </div>
  </div>,
  <div
    className="shrinked-view-container"
  >
    <div
      className="shrinked-view-icon-container"
    >
      <img
        className="shrinked-view-icon"
        height={20}
        src="../resources/missing-website-favicon.png"
        width={20}
      />
    </div>
    <div
      className="shrinked-view-time-spent-text-container"
    >
      <div
        className="shrinked-view-time-spent-text"
      >
        0
        :
        00
        :
        00
      </div>
    </div>
  </div>,
]
`;

exports[`Sorting renders correctly according to snapshot 1`] = `
<div
  className="sorting-container"
>
  <div
    className="sorting-options-label"
  >
    Sort: 
  </div>
  <div
    className="sorting-options-list-container"
  >
    <div
      className="sorting-option-tile-container"
      onClick={[Function]}
    >
      <div
        className="sorting-option-tile-label"
      >
        None
      </div>
    </div>
  </div>
</div>
`;

exports[`ViewChangeButton renders correctly regarding the snapshot 1`] = `
<button
  className="view-change-button"
  itemType="button"
  onClick={[Function]}
>
  <div
    className="view-change-button-text"
  >
    More
  </div>
</button>
`;
